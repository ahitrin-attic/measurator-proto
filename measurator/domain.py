import datetime


class IO:
    """Custom wrapper for all input/output interactions"""

    def write(self, text: str, *args) -> None:
        raise NotImplementedError

    def read(self) -> str:
        raise NotImplementedError

    def write_file(self, data: iter) -> None:
        raise NotImplementedError

    def read_file(self) -> iter:
        raise NotImplementedError

    def now(self) -> datetime.datetime:
        raise NotImplementedError
