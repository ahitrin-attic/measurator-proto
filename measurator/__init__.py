from measurator.main import run_main, migrate
