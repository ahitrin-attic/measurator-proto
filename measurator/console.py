import datetime

from measurator.domain import IO


class ConsoleIO(IO):
    def __init__(self, path) -> None:
        super().__init__()
        self._path = path

    def write(self, text: str, *args):
        print(text, *args)

    def read(self) -> str:
        return input()

    def write_file(self, data: iter) -> None:
        with open(self._path, "w") as f:
            for line in data:
                f.write(line)

    def read_file(self) -> iter:
        try:
            with open(self._path) as f:
                return f.readlines()
        except FileNotFoundError:
            return []

    def now(self) -> datetime.datetime:
        return datetime.datetime.now()
