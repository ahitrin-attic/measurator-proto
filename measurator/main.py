import argparse
import csv
import datetime
import time
from contextlib import ContextDecorator
from operator import itemgetter

from measurator.console import ConsoleIO
from measurator.domain import IO

TIME_FORMAT = "%Y-%m-%d %H:%M"


def migrate():
    pass


def run_main():
    args = _process_args()
    run_main_(ConsoleIO(args.path))


class FileWriteProxy(ContextDecorator):
    def __init__(self, io: IO):
        self.io = io
        self.cache = []

    def write(self, data) -> None:
        self.cache.append(data)

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        self.io.write_file(self.cache)


def run_main_(io: IO):
    not_yet, succeeds, fails = _read_file(io)
    # evaluate measurements
    delayed = list()
    for status, created, timestamp, text in not_yet:
        evaluate_time = datetime.datetime(*(time.strptime(timestamp, TIME_FORMAT)[:6]))
        if evaluate_time < io.now():
            io.write(f"Time to evaluate: {text}\n Is it true? (Delay/Reject/Yes/*No*)")
            user_input = io.read().capitalize()
            if user_input.startswith("Y"):
                status = "S"
                succeeds.append((status, created, timestamp, text))
            elif user_input.startswith("D"):
                io.write("When to evaluate (YYYY-mm-dd HH:MM):")
                eval_time = io.read()
                delayed.append(("N", created, eval_time, text))
            elif user_input.startswith('R'):
                io.write("Evaluation rejected")
            else:
                status = "F"
                fails.append((status, created, timestamp, text))
        else:
            delayed.append((status, created, timestamp, text))
    not_yet = delayed
    # print total statistics
    total_done = len(fails) + len(succeeds)
    if total_done > 0:
        percentage = "%d%%" % (float(100 * len(succeeds)) / float(total_done))
    else:
        percentage = "N/A"
    io.write("Successful predictions:", percentage, ", not done yet:", len(not_yet))
    # add another prediction when needed
    io.write("Add another prediction? Y/N")
    user_input = io.read()
    if user_input.capitalize().startswith("Y"):
        io.write("Prediction:")
        prediction = io.read()
        io.write("When to evaluate (YYYY-mm-dd HH:MM):")
        eval_time = io.read()
        not_yet.append(("N", io.now().strftime(TIME_FORMAT), eval_time, prediction))
    # overwrite predictions file
    with FileWriteProxy(io) as f:
        writer = csv.writer(f)
        for row in sorted(fails + succeeds + not_yet, key=itemgetter(1)):
            writer.writerow(row)


def _read_file(io: IO):
    fails = list()
    succeeds = list()
    not_yet = list()
    reader = csv.reader(io.read_file())
    for row in reader:
        status = row[0]
        if status == "F":
            fails.append(row)
        elif status == "S":
            succeeds.append(row)
        else:
            not_yet.append(row)
    return not_yet, succeeds, fails


def _process_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("path")
    args = parser.parse_args()
    return args
