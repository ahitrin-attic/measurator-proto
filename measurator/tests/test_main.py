import datetime
from typing import List

from approvaltests import verify
from approvaltests.reporters import GenericDiffReporterFactory

from measurator import run_main_
from measurator.domain import IO


class TestableIO(IO):
    def __init__(self, inputs: List[str], timestamps: List[datetime.datetime], file_content: iter):
        self.file_content = file_content
        self.inputs = inputs
        self.timestamps = timestamps
        self.log: List[str] = []

    def write(self, text: str, *args) -> None:
        self.log.append(f" {text}{''.join(str(x) for x in args)}")

    def read(self) -> str:
        output = self.inputs.pop()
        self.log.append(f"> {output}")
        return output

    def write_file(self, data: iter) -> None:
        self.log.append(f"! {'|'.join(data)}")

    def read_file(self) -> iter:
        self.log.append("* READ FILE")
        return self.file_content

    def now(self) -> datetime.datetime:
        timestamp = self.timestamps.pop()
        self.log.append(f"@ {timestamp}")
        return timestamp


def test_approve():
    inputs = ['N']
    timestamps = []
    file_content = []
    io = TestableIO(inputs, timestamps, file_content)
    run_main_(io)
    reporter = GenericDiffReporterFactory().get_first_working()
    verify("\n".join(io.log), reporter)
