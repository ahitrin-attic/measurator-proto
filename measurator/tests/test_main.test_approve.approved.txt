* READ FILE
 Successful predictions:N/A, not done yet:0
 Add another prediction? Y/N
> N
! 